File: Assets/Plugins/Android/com/adversityroad/web/WebScreen.java
package com.adversityroad.web;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.PixelFormat;
import android.net.Uri;
import android.util.DisplayMetrics;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.webkit.ValueCallback;
import android.webkit.WebChromeClient;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceResponse;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.FrameLayout;

import java.io.ByteArrayInputStream;
import java.util.HashMap;
import java.util.Map;

import com.unity3d.player.UnityPlayer;

/**
 * The picture of the in-game television: a web page drawn on top of the Unity view.
 * All calls come from C# (AndroidJavaClass.CallStatic).
 *
 * ASCII-only on purpose: javac's default encoding on the build machine is not
 * guaranteed to be UTF-8, and one non-ASCII byte is a hard build failure there.
 *
 * WHY AN OVERLAY AND NOT A TEXTURE
 * The only lawful way to show YouTube inside an app is a WebView (extracting the media
 * stream and feeding it to a video texture breaks YouTube's terms and rots within
 * weeks). A WebView renders into the Android view hierarchy, not into a GL texture we
 * could sample from a shader, so the picture has to be placed on top of the rectangle
 * the TV screen occupies on the display. C# projects the screen quad every frame and
 * calls place() only while that rectangle is standing still.
 *
 * WHY ITS OWN WINDOW (this is what fixed "sound plays but the picture is black")
 * Unity's activity is declared android:hardwareAccelerated="false" - Unity draws
 * through its own surface and does not need the view hierarchy accelerated. But an
 * HTML5 <video> element can only be composited on a hardware accelerated window; in a
 * software window Chromium plays the audio and leaves the video area black, which is
 * exactly what the player saw. An activity's window flag cannot be changed after the
 * window exists, so the WebView gets its own panel window instead, created with
 * FLAG_HARDWARE_ACCELERATED. That window is also NOT_TOUCHABLE, which is a cleaner way
 * to let every touch reach the game than intercepting them in a container view (the
 * overlay can cover most of the display; if it ate touches the player would lose the
 * joystick). Playback control therefore lives in the in-game panel.
 * If the panel window is refused for any reason, the code falls back to adding the
 * WebView to the activity's content view - the old behaviour, black video included.
 *
 * WHY park() INSTEAD OF setVisibility(GONE)
 * Chromium suspends media for a WebView that is not visible, so hiding the overlay that
 * way also killed the sound - including in picture in picture mode. Parking shrinks the
 * window to one pixel in the corner instead: still "visible" as far as the WebView is
 * concerned, so playback (and background audio) continues.
 */
public class WebScreen {

    /**
     * Two ways to get a YouTube video onto the screen, because YouTube refuses embeds
     * for several different reasons and only a device can tell which one applies. C#
     * starts with mode 0 and automatically retries with mode 1 when nothing plays a few
     * seconds later (see probe()).
     *
     * MODE 0 - our own player page served from a REAL https origin.
     *   The page is not fetched from a server: shouldInterceptRequest() answers the
     *   request for PAGE_URL with the HTML below, so the WebView believes the document
     *   really came from https://tv.adversityroad.app - an ordinary third party origin.
     *   Spoofing the origin with loadDataWithBaseURL("https://www.youtube.com", ...)
     *   does not work: an embed whose Referer is youtube.com itself is not a real embed
     *   and YouTube answers it with "this video cannot be watched here" (150 / 152).
     *
     * MODE 1 - load https://www.youtube.com/embed/<id> directly, WITH a Referer header.
     *   A WebView sends no Referer of its own, and YouTube answers a referer-less embed
     *   request with "error 153 - video player configuration error".
     *
     * Both modes drop the "; wv" marker from the User-Agent: that substring is how a
     * page can tell it is inside a WebView, and YouTube refuses part of its catalogue
     * when it sees it.
     */
    private static final String VIDEO = "document.querySelector('video')";

    /** Origin the player page claims to come from. Never actually fetched. */
    private static final String ORIGIN = "https://tv.adversityroad.app";
    private static final String PAGE_URL = ORIGIN + "/player.html";

    private static final String PAGE =
        "<!DOCTYPE html><html><head>" +
        "<meta name='viewport' content='width=device-width, initial-scale=1'>" +
        "<style>html,body{margin:0;padding:0;background:#000;overflow:hidden}" +
        "#p{width:100vw;height:100vh}</style></head><body><div id='p'></div>" +
        "<script src='https://www.youtube.com/iframe_api'></script><script>" +
        "var pl=null,arErr=0,want='__ID__',mute=__MUTE__;" +
        "function onYouTubeIframeAPIReady(){pl=new YT.Player('p',{videoId:want," +
        "playerVars:{autoplay:1,playsinline:1,rel:0,controls:0,modestbranding:1,fs:0," +
        "iv_load_policy:3,origin:'__ORIGIN__'}," +
        "events:{onReady:function(e){if(mute)e.target.mute();e.target.playVideo();}," +
        "onError:function(e){arErr=e.data;}," +
        "onStateChange:function(e){if(e.data==0)e.target.playVideo();}}});}" +
        "function arPlay(){if(pl)pl.playVideo();}" +
        "function arPause(){if(pl)pl.pauseVideo();}" +
        "function arMute(m){if(pl){if(m)pl.mute();else pl.unMute();}}" +
        "function arState(){try{if(arErr!=0)return '0:'+arErr;" +
        "var s=pl?pl.getPlayerState():-9;return (s==1||s==3)?'1':'0';}catch(e){return '0';}}" +
        "</script></body></html>";

    private static WebView sWeb;
    private static FrameLayout sFallbackHost;     // only used if the panel window fails
    private static WindowManager sWm;
    private static WindowManager.LayoutParams sLp;
    private static boolean sWindowMode;
    private static boolean sReady;
    private static int sMode;
    private static String sPageHtml = "";

    /** Container used only by the fallback path: steals touches, then declines them. */
    private static class PassThrough extends FrameLayout {
        PassThrough(Context c) { super(c); }
        @Override public boolean onInterceptTouchEvent(MotionEvent e) { return true; }
        @Override public boolean onTouchEvent(MotionEvent e) { return false; }
    }

    public static boolean available() {
        return UnityPlayer.currentActivity != null;
    }

    /** True when the picture is on its own hardware accelerated window (video works). */
    public static boolean accelerated() {
        return sWindowMode;
    }

    /**
     * Show / move the picture. x,y,w,h are in Unity's screen pixels (origin top left),
     * srcW/srcH is the Unity screen size they were measured against.
     *
     * The rectangle is rescaled here because Unity's surface and the Android display do
     * not have to be the same number of pixels (resolution scaling, a display cutout,
     * split screen, picture in picture). Placing Unity pixels straight into window
     * coordinates puts the picture next to the television instead of on it.
     */
    public static void place(final int x, final int y, final int w, final int h,
                             final int srcW, final int srcH) {
        final Activity a = UnityPlayer.currentActivity;
        if (a == null) return;
        a.runOnUiThread(new Runnable() {
            public void run() {
                if (!ensure(a)) return;
                float sx = 1f, sy = 1f;
                int refW = 0, refH = 0;
                if (sWindowMode) {
                    DisplayMetrics dm = new DisplayMetrics();
                    try {
                        a.getWindowManager().getDefaultDisplay().getRealMetrics(dm);
                        refW = dm.widthPixels;
                        refH = dm.heightPixels;
                    } catch (Throwable t) {
                        refW = refH = 0;
                    }
                } else {
                    View content = a.findViewById(android.R.id.content);
                    if (content != null) {
                        refW = content.getWidth();
                        refH = content.getHeight();
                    }
                }
                if (srcW > 0 && srcH > 0 && refW > 0 && refH > 0) {
                    sx = (float) refW / (float) srcW;
                    sy = (float) refH / (float) srcH;
                }
                int px = Math.round(x * sx), py = Math.round(y * sy);
                int pw = Math.max(4, Math.round(w * sx)), ph = Math.max(4, Math.round(h * sy));
                apply(a, px, py, pw, ph);
            }
        });
    }

    /**
     * Take the picture off the screen WITHOUT stopping playback: shrink it to a single
     * pixel in the corner. See the class comment for why this is not setVisibility(GONE).
     */
    public static void hide() {
        final Activity a = UnityPlayer.currentActivity;
        if (a == null) return;
        a.runOnUiThread(new Runnable() {
            public void run() {
                if (sWeb == null) return;
                apply(a, 0, 0, 1, 1);
            }
        });
    }

    /** Load a YouTube video by its 11 character id. mode: see the class comment. */
    public static void playYouTube(final String videoId, final boolean muted, final int mode) {
        final Activity a = UnityPlayer.currentActivity;
        if (a == null || videoId == null) return;
        a.runOnUiThread(new Runnable() {
            public void run() {
                if (!ensure(a)) return;
                sReady = false;
                sMode = mode == 1 ? 1 : 0;
                if (sMode == 0) {
                    sPageHtml = PAGE.replace("__ID__", videoId)
                                    .replace("__MUTE__", muted ? "true" : "false")
                                    .replace("__ORIGIN__", ORIGIN);
                    sWeb.loadUrl(PAGE_URL);
                    return;
                }
                // loop=1 needs playlist=<same id> - YouTube's own rule for looping a
                // single video. controls=0 because touches never reach the page anyway.
                String url = "https://www.youtube.com/embed/" + videoId
                    + "?autoplay=1&playsinline=1&rel=0&modestbranding=1&controls=0"
                    + "&iv_load_policy=3&loop=1&playlist=" + videoId
                    + (muted ? "&mute=1" : "");
                Map<String, String> headers = new HashMap<String, String>();
                headers.put("Referer", "https://www.youtube.com/");
                sWeb.loadUrl(url, headers);
            }
        });
    }

    /** Load any http(s) page directly (a plain .mp4 link, a live stream page...). */
    public static void loadUrl(final String url) {
        final Activity a = UnityPlayer.currentActivity;
        if (a == null || url == null) return;
        a.runOnUiThread(new Runnable() {
            public void run() {
                if (!ensure(a)) return;
                sReady = false;
                sMode = 1;
                sWeb.loadUrl(url);
            }
        });
    }

    public static void play() {
        post(sMode == 0 ? "arPlay()" : "var v=" + VIDEO + ";if(v)v.play();");
    }

    public static void pause() {
        post(sMode == 0 ? "arPause()" : "var v=" + VIDEO + ";if(v)v.pause();");
    }

    public static void mute(final boolean m) {
        post(sMode == 0 ? ("arMute(" + (m ? "true" : "false") + ")")
                        : ("var v=" + VIDEO + ";if(v)v.muted=" + (m ? "true" : "false") + ";"));
    }

    /**
     * Ask the page whether anything is actually playing and report back to Unity via
     * UnitySendMessage(callbackObject, "OnWebPlayback", "1" / "0" / "0:<errorCode>").
     *
     * Some videos forbid embedding, some refuse the origin; the page then shows
     * YouTube's own error card and nothing plays. Without this the game cannot tell
     * that apart from "still buffering" and the player is left staring at a box.
     */
    public static void probe(final String callbackObject) {
        final Activity a = UnityPlayer.currentActivity;
        if (a == null || callbackObject == null) return;
        a.runOnUiThread(new Runnable() {
            public void run() {
                if (sWeb == null) {
                    send(callbackObject, "0");
                    return;
                }
                try {
                    String q = sMode == 0
                        ? "arState()"
                        : "(function(){var v=" + VIDEO
                          + ";return (v&&v.readyState>0&&!v.paused)?'1':'0';})()";
                    sWeb.evaluateJavascript(q, new ValueCallback<String>() {
                        public void onReceiveValue(String value) {
                            send(callbackObject, value == null ? "0" : value.replace("\"", ""));
                        }
                    });
                } catch (Throwable t) {
                    send(callbackObject, "0");
                }
            }
        });
    }

    /** Keep the page running while the app is in the background / in a floating window. */
    public static void keepAlive() {
        final Activity a = UnityPlayer.currentActivity;
        if (a == null) return;
        a.runOnUiThread(new Runnable() {
            public void run() {
                if (sWeb == null) return;
                sWeb.resumeTimers();
                sWeb.onResume();
            }
        });
    }

    /** Tear the picture down completely (leaving home, shutting the set off). */
    public static void close() {
        final Activity a = UnityPlayer.currentActivity;
        if (a == null) return;
        a.runOnUiThread(new Runnable() {
            public void run() {
                try {
                    if (sWeb != null) {
                        sWeb.loadUrl("about:blank");
                        sWeb.stopLoading();
                        if (sWindowMode && sWm != null) {
                            sWm.removeViewImmediate(sWeb);
                        } else {
                            ViewGroup parent = sFallbackHost != null
                                ? (ViewGroup) sFallbackHost.getParent() : null;
                            if (parent != null) parent.removeView(sFallbackHost);
                        }
                        sWeb.destroy();
                    }
                } catch (Throwable t) {
                    // already detached
                }
                sWeb = null;
                sFallbackHost = null;
                sWm = null;
                sLp = null;
                sWindowMode = false;
                sReady = false;
                sPageHtml = "";
            }
        });
    }

    /** Hand the link to whatever app the player has (YouTube app, browser). */
    public static void openExternal(String url) {
        Activity a = UnityPlayer.currentActivity;
        if (a == null || url == null) return;
        try {
            Intent i = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
            i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            a.startActivity(i);
        } catch (Throwable t) {
            // no app can open it - the in-game panel already said what is playing
        }
    }

    // ================= internals (UI thread only) =================

    private static void apply(Activity a, int x, int y, int w, int h) {
        try {
            if (sWindowMode && sWm != null && sLp != null) {
                sLp.x = x;
                sLp.y = y;
                sLp.width = w;
                sLp.height = h;
                sWm.updateViewLayout(sWeb, sLp);
                return;
            }
            if (sFallbackHost != null) {
                FrameLayout.LayoutParams lp = new FrameLayout.LayoutParams(w, h);
                lp.leftMargin = x;
                lp.topMargin = y;
                sFallbackHost.setLayoutParams(lp);
                sFallbackHost.setVisibility(View.VISIBLE);
                sFallbackHost.requestLayout();
            }
        } catch (Throwable t) {
            // the window went away under us; the next place() rebuilds it
        }
    }

    private static boolean ensure(Activity a) {
        if (sWeb != null) return true;
        try {
            sWeb = new WebView(a);
            WebSettings s = sWeb.getSettings();
            s.setJavaScriptEnabled(true);
            s.setDomStorageEnabled(true);
            // Without this the page needs a tap before it may play - and taps never
            // reach the page, so the TV would stay black forever.
            s.setMediaPlaybackRequiresUserGesture(false);
            s.setUseWideViewPort(true);
            s.setLoadWithOverviewMode(true);
            String ua = s.getUserAgentString();
            if (ua != null && ua.contains("; wv")) s.setUserAgentString(ua.replace("; wv", ""));

            sWeb.setBackgroundColor(Color.BLACK);
            sWeb.setLayerType(View.LAYER_TYPE_HARDWARE, null);
            sWeb.setWebChromeClient(new WebChromeClient());
            sWeb.setWebViewClient(new WebViewClient() {
                // Serve our own player page for PAGE_URL without any network request.
                // This is what gives the document a real (and stable) origin.
                @Override
                public WebResourceResponse shouldInterceptRequest(WebView v, WebResourceRequest req) {
                    try {
                        if (req != null && req.getUrl() != null
                                && PAGE_URL.equals(req.getUrl().toString())
                                && sPageHtml.length() > 0) {
                            return new WebResourceResponse("text/html", "utf-8",
                                new ByteArrayInputStream(sPageHtml.getBytes("UTF-8")));
                        }
                    } catch (Throwable t) {
                        // fall through and let the WebView try the network
                    }
                    return null;
                }

                @Override public void onPageFinished(WebView v, String url) {
                    sReady = true;
                }
            });

            if (attachWindow(a)) return true;
            return attachFallback(a);
        } catch (Throwable t) {
            sWeb = null;
            return false;
        }
    }

    /** The good path: a hardware accelerated, untouchable panel window. */
    private static boolean attachWindow(Activity a) {
        try {
            View decor = a.getWindow() != null ? a.getWindow().getDecorView() : null;
            if (decor == null || decor.getWindowToken() == null) return false;

            sWm = a.getWindowManager();
            sLp = new WindowManager.LayoutParams(
                1, 1,
                WindowManager.LayoutParams.TYPE_APPLICATION_PANEL,
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                    | WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE
                    | WindowManager.LayoutParams.FLAG_HARDWARE_ACCELERATED
                    | WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
                PixelFormat.TRANSLUCENT);
            sLp.gravity = Gravity.TOP | Gravity.LEFT;
            sLp.x = 0;
            sLp.y = 0;
            sLp.token = decor.getWindowToken();
            sWm.addView(sWeb, sLp);
            sWindowMode = true;
            return true;
        } catch (Throwable t) {
            sWm = null;
            sLp = null;
            sWindowMode = false;
            return false;
        }
    }

    /** The old path: a view inside the activity's content view (video will be black). */
    private static boolean attachFallback(Activity a) {
        try {
            sWeb.setLayoutParams(new FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
            sFallbackHost = new PassThrough(a);
            sFallbackHost.setBackgroundColor(Color.BLACK);
            sFallbackHost.addView(sWeb);
            a.addContentView(sFallbackHost, new FrameLayout.LayoutParams(1, 1));
            return true;
        } catch (Throwable t) {
            sWeb = null;
            sFallbackHost = null;
            return false;
        }
    }

    private static void post(final String code) {
        final Activity a = UnityPlayer.currentActivity;
        if (a == null) return;
        a.runOnUiThread(new Runnable() {
            public void run() { js(code); }
        });
    }

    private static void js(String code) {
        if (sWeb == null) return;
        try {
            sWeb.evaluateJavascript(code, null);
        } catch (Throwable t) {
            // page not loaded yet; the next call will land
        }
    }

    private static void send(String obj, String value) {
        try {
            UnityPlayer.UnitySendMessage(obj, "OnWebPlayback", value);
        } catch (Throwable t) {
            // the bridge object is gone; nothing to report to
        }
    }
}

File: Assets/_Project/Scripts/Platform/WebScreen.cs
using UnityEngine;

namespace AdversityRoad.Platform
{
    /// <summary>
    /// 安卓网页层：贴在 Unity 画面之上的一块网页，用来当游戏里那台电视的**画面**。
    ///
    /// 【为什么是"贴在画面上"而不是贴图】
    /// 想在应用内看 YouTube，唯一正路是 WebView：把视频流解出来喂给 VideoPlayer
    /// 既违反 YouTube 条款，也每隔几周就失效一次。而 WebView 渲染进的是安卓的 View
    /// 树，不是我们能采样的 GL 贴图——所以画面只能是一块**盖在电视屏幕所在矩形上**
    /// 的原生视图：C# 每帧把屏幕四角投影到屏幕空间，算出包围矩形交给它；玩家不在
    /// 电视正面时收起网页层，改由世界里的屏幕自己显示程序生成的画面（见 WallTv）。
    ///
    /// 【触摸一律穿透】网页层可能盖住大半个屏幕（站在 5 米宽的电视前时），
    /// 要是它吃掉触摸，玩家就没法操作摇杆了。Java 侧的容器会先截走触摸再拒绝它，
    /// 于是事件继续派发给下面的 Unity 视图。所有播放控制走游戏内面板（本类的方法）。
    /// </summary>
    public static class WebScreen
    {
        const string JavaClass = "com.adversityroad.web.WebScreen";

        static AndroidJavaClass _cls;
        static bool _probed, _ok;

        /// <summary>这台设备上能不能用网页层（只有安卓真机能）。</summary>
        public static bool Available
        {
            get
            {
                if (_probed) return _ok;
                _probed = true;
#if UNITY_ANDROID && !UNITY_EDITOR
                try
                {
                    _cls = new AndroidJavaClass(JavaClass);
                    _ok = _cls != null && _cls.CallStatic<bool>("available");
                }
                catch (System.Exception e)
                {
                    Debug.LogWarning("[WebScreen] 网页层不可用：" + e.Message);
                    _cls = null;
                    _ok = false;
                }
#else
                _ok = false;
#endif
                return _ok;
            }
        }

        /// <summary>
        /// 把网页层摆到屏幕上的这个矩形里（Unity 屏幕坐标，原点在左下）。
        /// 安卓的视图坐标原点在左上，这里负责翻过来；
        /// 同时把"这个矩形是按多大的 Unity 画面量出来的"一起传过去——
        /// Unity 的画面像素数和安卓视图的像素数**不一定相等**（分辨率缩放、刘海、
        /// 分屏、悬浮小窗都会让两者对不上），Java 侧按这个比例换算，
        /// 否则画面会摆到电视旁边去，还会随着差值缩放。
        /// </summary>
        public static void Place(Rect r)
        {
            if (!Available) return;
            int x = Mathf.RoundToInt(r.xMin);
            int y = Mathf.RoundToInt(Screen.height - r.yMax);
            int w = Mathf.RoundToInt(r.width);
            int h = Mathf.RoundToInt(r.height);
            Call("place", x, y, w, h, Screen.width, Screen.height);
        }

        /// <summary>收起画面，但**不停播**——走开时声音继续，才叫"后台播放"。</summary>
        public static void Hide() => Call("hide");

        /// <summary>
        /// 放一条 YouTube 视频。
        /// mode 0 = 自建 IFrame API 页面（默认，最通用）；
        /// mode 1 = 直接加载 embed 页并补上 Referer（对付"错误 153 播放器配置错误"）。
        /// 两条路都可能被个别视频拒绝，所以由调用方在探测失败后换一条重试。
        /// </summary>
        public static void PlayYouTube(string videoId, bool muted, int mode = 0)
        {
            if (string.IsNullOrEmpty(videoId)) return;
            Call("playYouTube", videoId, muted, mode);
        }

        public static void LoadUrl(string url) => Call("loadUrl", url);

        /// <summary>
        /// 查一下"到底有没有在播"，结果通过 <see cref="PlaybackChecked"/> 回来。
        /// 有些视频不许嵌入播放，页面上会出现 YouTube 自己的"此视频不能观看"卡片，
        /// 而游戏这边看起来和"还在缓冲"一模一样——玩家只能对着一个黑框发呆。
        /// </summary>
        public static void Probe()
        {
            if (!Available) return;
            EnsureBridge();
            Call("probe", BridgeName);
        }

        /// <summary>Probe 的结果：true = 页面里真的有个视频在播。</summary>
        public static event System.Action<bool> PlaybackChecked;
        public static void Play() => Call("play");
        public static void Pause() => Call("pause");
        public static void Mute(bool m) => Call("mute", m);
        public static void KeepAlive() => Call("keepAlive");
        public static void Close() => Call("close");
        public static void OpenExternal(string url) => Call("openExternal", url);

        const string BridgeName = "WebScreenBridge";
        static Bridge _bridge;

        static void EnsureBridge()
        {
            if (_bridge != null) return;
            var go = new GameObject(BridgeName);
            Object.DontDestroyOnLoad(go);
            _bridge = go.AddComponent<Bridge>();
        }

        /// <summary>接 Java 侧 UnitySendMessage 的固定收件人。</summary>
        class Bridge : MonoBehaviour
        {
            public void OnWebPlayback(string v) => PlaybackChecked?.Invoke(v == "1");
        }

        static void Call(string method, params object[] args)
        {
            if (!Available || _cls == null) return;
            try { _cls.CallStatic(method, args); }
            catch (System.Exception e) { Debug.LogWarning("[WebScreen] " + method + " 失败：" + e.Message); }
        }

        /// <summary>
        /// 从玩家粘进来的东西里抠出 YouTube 视频 id。
        /// 支持 watch?v= / youtu.be/ / /shorts/ / /embed/ / 直接就是 11 位 id。
        /// 抠不出来返回空串（调用方会把它当普通网址处理）。
        /// </summary>
        public static string ParseVideoId(string s)
        {
            if (string.IsNullOrEmpty(s)) return "";
            s = s.Trim();
            if (IsId(s)) return s;

            string[] keys = { "v=", "youtu.be/", "/shorts/", "/embed/", "/live/" };
            foreach (var k in keys)
            {
                int i = s.IndexOf(k, System.StringComparison.OrdinalIgnoreCase);
                if (i < 0) continue;
                string tail = s.Substring(i + k.Length);
                int cut = tail.IndexOfAny(new[] { '&', '?', '/', '#' });
                if (cut >= 0) tail = tail.Substring(0, cut);
                if (IsId(tail)) return tail;
            }
            return "";
        }

        /// <summary>id 只允许 11 位的 [A-Za-z0-9_-]：既是 YouTube 的格式，
        /// 也顺手挡掉往页面里塞脚本的可能（id 是直接拼进 JS 的）。</summary>
        public static bool IsId(string s)
        {
            if (string.IsNullOrEmpty(s) || s.Length != 11) return false;
            foreach (char c in s)
            {
                bool ok = (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
                       || (c >= '0' && c <= '9') || c == '_' || c == '-';
                if (!ok) return false;
            }
            return true;
        }
    }
}
